[Swift in WebKit] Add a Swift module to bmalloc to better support Swift on Cocoa platforms

rr-codes opened the pull request:
#### 14c2ad6d632057efcb311c363cb2f5cb685b8f78
<pre>
[Swift in WebKit] Add a Swift module to bmalloc to better support Swift on Cocoa platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=301054">https://bugs.webkit.org/show_bug.cgi?id=301054</a>
<a href="https://rdar.apple.com/162959248">rdar://162959248</a>

Reviewed by Elliott Williams and Mike Wyrzykowski.

Enable Swift and Swift-Cxx interop in bmalloc. In addition to enabling the possibility of using
Swift directly in bmalloc, it serves two important immediate-term purposes:

1. The module verifier tool does not currently support libraries, such as WTF and bmalloc. Compiling a
Swift file in bmalloc itself allows for at least a minimal means of verification (specifically, successful
compilation of this file indicates a necessary but not sufficient condition).

2. Enabling interop in bmalloc is an effective stepping stone to more complex projects like WTF
and WebCore, particularly because it is lower on the WebKit stack.

* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap:
* Source/bmalloc/bmalloc.swift: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/302570@main">https://commits.webkit.org/302570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/063c3185284ef914e6c1e44f47bd50ae8c6c4993

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135518 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79645 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0ad1a032-d105-4c39-bf22-8635cf7d5a7c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/363 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97555 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65449 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114806 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78124 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d572558c-65f9-4e9a-bd11-81c63bcfef66) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/214 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32913 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78830 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120189 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33398 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138009 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126618 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/292 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/273 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106084 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/324 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111148 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105822 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27250 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/224 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29702 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52525 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/338 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159636 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/245 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39853 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/317 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/297 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
```diff
@@ -21,6 +21,12 @@
 /* End PBXAggregateTarget section */
 
 /* Begin PBXBuildFile section */
+		0100000C37BABA0A0A993999 /* pas_mte_config.c in Sources */ = {isa = PBXBuildFile; fileRef = 0100000A37BABA0A0A993999 /* pas_mte_config.c */; };
```
Xcode wanted to sort these files so I let it

adetaylor commented on Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
```diff
@@ -21,6 +21,12 @@
+		0100000D37BABA0A0A991999 /* pas_zero_memory.h in Headers */ = {isa = PBXBuildFile; fileRef = 0100000B37BABA0A0A991999 /* pas_zero_memory.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		0100000D37BABA0A0A993999 /* pas_mte_config.h in Headers */ = {isa = PBXBuildFile; fileRef = 0100000B37BABA0A0A993999 /* pas_mte_config.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		0100000D37BABA0A0A999999 /* pas_mte.h in Headers */ = {isa = PBXBuildFile; fileRef = 0100000B37BABA0A0A999999 /* pas_mte.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		07FB5E3D2EA4B31B00603B46 /* bmalloc.swift in Sources */ = {isa = PBXBuildFile; fileRef = 07FB5E3C2EA4B31B00603B46 /* bmalloc.swift */; };
```
I'm really surprised you don't have to make this conditional at least on `GENERATE_BMALLOC_MODULEMAP`

rr-codes replied:
GENERATE_BMALLOC_MODULEMAP isn’t actually used anymore (I just forgot to remove it)

adetaylor replied:
It's used in `bmalloc.xcconfig` in `EXCLUDED_SOURCE_FILE_NAMES` right? (Until this PR - if you're intentionally removing it that's fine with me)

emw-apple commented on Source/bmalloc/Configurations/bmalloc.xcconfig:
```diff
@@ -54,3 +54,19 @@ STRIP_INSTALLED_PRODUCT = NO;
 // is not exported to script phases, so it must be bound to a different name.
 WK_HEADERS_VFS_FILE = $(CPP_HEADERMAP_PRODUCT_HEADERS_VFS_FILE);
+
+INSTALLHDRS_SCRIPT_PHASE = YES;
```
Why? This PR doesn't add any script phases.

emw-apple reviewed (approved):
What kind of impact on build time does this have? In theory it should be minimal, but want to be sure we're not paying too high a cost.

rr-codes committed:
[Swift in WebKit] Add a Swift module to bmalloc to better support Swift on Cocoa platforms
https://bugs.webkit.org/show_bug.cgi?id=301054
rdar://162959248

Reviewed by Elliott Williams and Mike Wyrzykowski.

Enable Swift and Swift-Cxx interop in bmalloc. In addition to enabling the possibility of using
Swift directly in bmalloc, it serves two important immediate-term purposes:

1. The module verifier tool does not currently support libraries, such as WTF and bmalloc. Compiling a
Swift file in bmalloc itself allows for at least a minimal means of verification (specifically, successful
compilation of this file indicates a necessary but not sufficient condition).

2. Enabling interop in bmalloc is an effective stepping stone to more complex projects like WTF
and WebCore, particularly because it is lower on the WebKit stack.

* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap:
* Source/bmalloc/bmalloc.swift: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/302570@main
